### PR TITLE
Enabling pre-push checks on Windows

### PR DESCRIPTION
**Pull request by mtolmacs**
Re-submit the changes to enable windows pre-push checks.

This patch changes how `ci/bin/format.dart` generate diffs from `diff` and `patch` commands to `git diff` and `git apply` in order to have a common method for these operations on all platforms. Windows installations don't have diff and patch commands available by default and many implementations which provide such commands work differently than the UN*X tools. Git however works consistently across all platforms.

Additionally, this patch also changes the python executable in some of the pre-push components affected by this to `vpython3` to continue the effort started at flutter/flutter#108474 and I also removed the `--no-sound-null-safety` parameter in the ci/format.sh, ci/format.bat files

NOTE: Since the original patch caused some issues, I suggest that this should be tested more carefully before it is merged.

### Issues fixed by this PR
* flutter/flutter#108122
* flutter/flutter#107920
* flutter/flutter#86506
* flutter/flutter#106615

### [flutter/tests] repo impact
None.

## Pre-launch Checklist

- [X] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [X] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [X] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [X] I listed at least one issue that this PR fixes in the description above.
- [X] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt. See [testing the engine] for instructions on
writing and running engine tests.
- [X] I updated/added relevant documentation (doc comments with `///`).
- [X] I signed the [CLA].
- [X] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/main/CONTRIBUTING.md#style
[testing the engine]: https://github.com/flutter/flutter/wiki/Testing-the-engine
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

File: ci/bin/format.dart [hunk 1]
```diff
@@ -226,7 +225,7 @@ abstract class FormatChecker {
     );
     final List<WorkerJob> jobs = patches.map<WorkerJob>((String patch) {
       return WorkerJob(
-        <String>['patch', '-p0'],
+        <String>['git', 'apply', '--ignore-space-change'],
```
Comment (loic-sharma) [inline ci/bin/format.dart:228]: From my understanding, `--ignore-space-change` might ignore real problems. Should we only add `--ignore-space-change` if on `Platform.isWindows`?

File: ci/format.bat [hunk 2]
```diff
@@ -28,5 +28,4 @@ cd "%ci_dir%"
 REM Do not use the CALL command in the next line to execute Dart. CALL causes
 REM Windows to re-read the line from disk after the CALL command has finished
 REM regardless of the ampersand chain.
-REM TODO(gspencergoog): Remove --no-sound-null-safety once isolate package is null-safe.
-"%dart%" --no-sound-null-safety --disable-dart-dev bin\format.dart %* & exit /B !ERRORLEVEL!
+"%dart%" --disable-dart-dev bin\format.dart %* & exit /B !ERRORLEVEL!
```
Comment (cbracken) [inline ci/format.bat:31]: It may be worth moving the null safety switch to a separate PR so that if we have to revert one we don't need to revert the other.

**Comment by loic-sharma** [comment]
I might be missing something, but should this PR update the githooks path?

https://github.com/flutter/engine/blob/dbb5284f2c8d87473d60baa21bba0b7ab88c8281/tools/githooks/setup.py#L30-L32

File: ci/bin/format.dart (hunk 1 above)
Comment (loic-sharma) [reply ci/bin/format.dart:228]: From my understanding this fixes formatting issues, and `--ignore-space-change` is to skip line ending changes on Windows. If formatting replaced tabs with spaces, would that get ignored here?

> I'm asking because if those problems are present on Windows as well, then making it conditional does not solve the problem, only limits the damage it can done. I'd much rather solve it proper, if possible.

I completely agree 👍

**Comment by loic-sharma** [comment]
When I test this locally, I get a pre-push warning even after I run `gclient sync -D`:

```
> gh co 36123
> git push -u origin HEAD
Warning: Pre-push checks are not enabled!
Note: Please run 'gclient sync -D' to enable pre-push checks on Windows
...
> gclient sync -D
Updating depot_tools...
...
Running hooks: 100% (10/10), done.
> git push origin HEAD
Warning: Pre-push checks are not enabled!
Note: Please run 'gclient sync -D' to enable pre-push checks on Windows
Everything up-to-date
```

I'm a bit confused. Does this change intend to keep a Windows pre-push hook to kick off formatting? If so, it looks like the current Windows pre-push hook is incorrect (it outputs a warning instead of kicking off formatting). Or does this change intend to have a single pre-push hook for all platforms? If so, it seems `setup.py` should be updated. Please let me know if I misunderstood something.

File: tools/githooks/pre-push [hunk 3]
```diff
@@ -1,4 +1,4 @@
-#!/usr/bin/env python3
+#!/usr/bin/env vpython3
```
Comment (mtolmacs) [inline tools/githooks/pre-push:1]: Avoids issue with Windows setting up alias for python3 to the Microsoft Store by default. This will also make sure that every install uses the depot_tools python version.

**Comment by mtolmacs** [comment]
@loic-sharma Haha you're right, I confused myself along the way and this was missing. I added it now.

File: ci/format.bat (hunk 2 above)
Comment (mtolmacs) [reply ci/format.bat]: It seems that this was changed in main separately from this PR already.

**Comment by loic-sharma** [comment]
@cbracken @yaakovschectman would you mind testing that this works as expected on your Windows machine? 

1. Modify a file locally such that its formatting is incorrect
2. Checkout this branch:
```
> git checkout -b mtolmacs-reenable-win-prepush main
> git pull git@github.com:mtolmacs/flutter-engine.git reenable-win-prepush
```

3. Check you get a warning if the pre-push hook path hasn't been updated yet:

```
> git push -u origin HEAD
Warning: Pre-push checks are not enabled!
Note: Please run 'gclient sync -D' to enable pre-push checks on Windows
...
```

4. Update your pre-push hook

```
> gclient sync -D
Updating depot_tools...
...
Running hooks: 100% (10/10), done.
```

5. Verify it catches formatting issues:

```
> git push origin HEAD
The clang-tidy check is disabled. To enable set the environment variable PRE_PUSH_CLANG_TIDY to any value.
Starting formatting checks.
Check "Formatting check" failed.
...
formatting checks finished in 0:00:02.547010
pre-push checks finished in 0:00:02.555945
error: failed to push some refs to 'github.com:loic-sharma/flutter-engine.git'
```

File: ci/bin/format.dart (hunk 1 above)
Comment (cbracken) [reply ci/bin/format.dart:228]: I wonder if on Windows only we could enable `--ignore-cr-at-eol` if that's currently causing an issue with the formatter? I assume that on push, we're normalising line endings to unix line endings?

**Review by cbracken — COMMENTED** [review]
I'll give a test on my Windows machine. Have we re-tested everything looks good for non-Windows too?

File: ci/bin/format.dart (hunk 1 above)
Comment (mtolmacs) [reply ci/bin/format.dart:228]: The issue this switch is trying to solve comes from `clang-tidy` on Windows. No matter what line endings you feed into it, the output is always CRLF EOF. So [I use `--ignore-cr-at-eol` at patch creation](https://github.com/flutter/engine/pull/36123/files/8a5be011020666c371a8cacb7bba950c0b0be849#diff-b12ae99b59c24a1d8d53205e819e46da40d4d3869a1cdf509a7905450e09ce38R408) in order to skip patching line endings to CRLF for the whole file. 

As for the patch application side, since `--ignore-cr-at-eol` is a `git diff`-only switch, I needed to find some other solution to avoid line ending differences tripping the context matching in `git apply`. If I remove this switch now, git apply says the patch does not apply (obviously, since the context lines in the patch differ in EOL).

I reviewed the [`--ignore-space-change` switch documentation](https://git-scm.com/docs/git-apply#Documentation/git-apply.txt---ignore-space-change) and I still can't see where this can cause a problem. The switch only controls how the application context is being checked, it does not change the patch itself. So if `clang-tidy` generates a properly formatted file, `git diff` will create the proper patch. Thus git apply (I think) cannot "misapply" the patch if the file did not change since running `format`.

Alternatively I can patch the output of `clang-tidy` EOL, but that could become complicated as it's not guaranteed that files on Windows are checked out with CRLF. You can easily check out files with the proper LF EOL, so that needs to be detected before patching the `clang-tidy` output.

**Comment by mtolmacs** [comment]
Added tests for `ci/bin/format.dart` in [460ad05](https://github.com/flutter/engine/pull/36123/commits/460ad0522be4e5c4ae689d48ba0deaa2920c88e3). Unfortunately the tests need to run in a serialized manner due to the fact that the formatting process cannot run in parallel. I solved it by using the 'sync' versions of dart:io commands, which is not ideal, but gets the job done.

One way forward could be to refactor `ci/bin/format.dart` to make it testing friendly, which might be out of scope for this PR. Let me know if you see a better approach and I'm happy to rewrite accordingly.

**Comment by loic-sharma** [comment]
> Have we re-tested everything looks good for non-Windows too?

Yup this works fine on my mac & on WSL

**Review by loic-sharma — APPROVED** [review]

File: ci/bin/format.dart (hunk 1 above)
Comment (cbracken) [reply ci/bin/format.dart:228]: Ah I see! Makes sense. Not sure why I misread that as `git diff` when it was clearly the patch command, probably mixing it up with the diff bits later. Thanks for clarifying.

**Review by cbracken — APPROVED** [review]
LGTM stamp from a Japanese personal seal

**Comment by loic-sharma** [comment]
@mtolmacs I rebase this on the latest `main` branch and got some incorrect formatting errors:

<details>
<summary>Incorrect formatting...</summary>


```
PS C:\Code\f\engine\src\flutter> git push
The clang-tidy check is disabled. To enable set the environment variable PRE_PUSH_CLANG_TIDY to any value.
Starting formatting checks.
Check "Formatting check" failed.
command: C:\Code\f\engine\src\flutter\ci\format.bat
working directory: C:\Code\f\engine\src\flutter
exit code: 1
stdout:
To fix, run:

git apply <<DONE
diff --git a/impeller/entity/contents/color_source_contents.cc b/impeller/entity/contents/color_source_contents.cc
index c95507ea4f..0000000000 100644
--- a/impeller/entity/contents/color_source_contents.cc
+++ b/impeller/entity/contents/color_source_contents.cc
@@ -38,17 +38,4 @@ const Matrix& ColorSourceContents::GetInverseMatrix() const {
 }

 std::optional<Rect> ColorSourceContents::GetCoverage(
-    const Entity& entity) const {
-  return geometry_->GetCoverage(entity.GetTransformation());
-};
-
-bool ColorSourceContents::ShouldRender(
-    const Entity& entity,
-    const std::optional<Rect>& stencil_coverage) const {
-  if (!stencil_coverage.has_value()) {
-    return false;
-  }
-  return Contents::ShouldRender(entity, stencil_coverage);
-}
-
-}  // namespace impeller
+    const Entity& entity)
\ No newline at end of file
diff --git a/impeller/entity/contents/color_source_contents.h b/impeller/entity/contents/color_source_contents.h
index 4396c38395..0000000000 100644
--- a/impeller/entity/contents/color_source_contents.h
+++ b/impeller/entity/contents/color_source_contents.h
@@ -39,11 +39,4 @@ class ColorSourceContents : public Contents {
   Scalar GetAlpha() const;

  private:
-  std::shared_ptr<Geometry> geometry_;
-  Matrix inverse_matrix_;
-  Scalar alpha_ = 1.0;
-
-  FML_DISALLOW_COPY_AND_ASSIGN(ColorSourceContents);
-};
-
-}  // namespace impeller
+  std::share
\ No newline at end of file
diff --git a/impeller/playground/compute_playground_test.h b/impeller/playground/compute_playground_test.h
index 06c480b5be..0000000000 100644
--- a/impeller/playground/compute_playground_test.h
+++ b/impeller/playground/compute_playground_test.h
@@ -39,13 +39,4 @@ class ComputePlaygroundTest
  private:
   fml::TimeDelta start_time_;

-  FML_DISALLOW_COPY_AND_ASSIGN(ComputePlaygroundTest);
-};
-
-#define INSTANTIATE_COMPUTE_SUITE(playground)                              \
-  INSTANTIATE_TEST_SUITE_P(                                                \
-      Compute, playground, ::testing::Values(PlaygroundBackend::kMetal),   \
-      [](const ::testing::TestParamInfo<ComputePlaygroundTest::ParamType>& \
-             info) { return PlaygroundBackendToString(info.param); });
-
-}  // namespace impeller
+  FML_DISALLOW_COPY_AND_ASSIGN(ComputePlaygroundTes
\ No newline at end of file
diff --git a/impeller/scene/importer/conversions.h b/impeller/scene/importer/conversions.h
index cb11d3c42f..0000000000 100644
--- a/impeller/scene/importer/conversions.h
+++ b/impeller/scene/importer/conversions.h
@@ -41,14 +41,4 @@ std::unique_ptr<fb::Matrix> ToFBMatrixUniquePtr(const Matrix& m);

 fb::Vec2 ToFBVec2(const Vector2 v);

-fb::Vec3 ToFBVec3(const Vector3 v);
-
-fb::Vec4 ToFBVec4(const Vector4 v);
-
-fb::Color ToFBColor(const Color c);
-
-std::unique_ptr<fb::Color> ToFBColor(const std::vector<double>& c);
-
-}  // namespace importer
-}  // namespace scene
-}  // namespace impeller
+fb::Vec3 ToFBVec3(const Ve
\ No newline at end of file
diff --git a/shell/platform/windows/testing/flutter_windows_engine_builder.h b/shell/platform/windows/testing/flutter_windows_engine_builder.h
index cc41f03fee..0000000000 100644
--- a/shell/platform/windows/testing/flutter_windows_engine_builder.h
+++ b/shell/platform/windows/testing/flutter_windows_engine_builder.h
@@ -27,17 +27,4 @@ class FlutterWindowsEngineBuilder {

   // Prevent copying.
   FlutterWindowsEngineBuilder(FlutterWindowsEngineBuilder const&) = delete;
-  FlutterWindowsEngineBuilder& operator=(FlutterWindowsEngineBuilder const&) =
-      delete;
-
- private:
-  WindowsTestContext& context_;
-  FlutterDesktopEngineProperties properties_ = {};
-  std::string dart_entrypoint_;
-  std::vector<std::string> dart_entrypoint_arguments_;
-};
-
-}  // namespace testing
-}  // namespace flutter
-
-#endif  // FLUTTER_SHELL_PLATFORM_WINDOWS_TESTING_FLUTTER_WINDOWS_ENGINE_BUILDER_H_
+  FlutterWindowsEngineBuilde
\ No newline at end of file
diff --git a/third_party/accessibility/ax/platform/ax_platform_tree_manager.h b/third_party/accessibility/ax/platform/ax_platform_tree_manager.h
index e9ee973ce2..0000000000 100644
--- a/third_party/accessibility/ax/platform/ax_platform_tree_manager.h
+++ b/third_party/accessibility/ax/platform/ax_platform_tree_manager.h
@@ -27,13 +27,4 @@ class AX_EXPORT AXPlatformTreeManager : public AXTreeManager {
       const AXNode::AXID node_id) const = 0;

   // Returns an AXPlatformNode that corresponds to the given |node|.
-  virtual AXPlatformNode* GetPlatformNodeFromTree(const AXNode& node) const = 0;
-
-  // Returns an AXPlatformNodeDelegate that corresponds to a root node
-  // of the accessibility tree.
-  virtual AXPlatformNodeDelegate* RootDelegate() const = 0;
-};
-
-}  // namespace ui
-
-#endif  // UI_ACCESSIBILITY_PLATFORM_AX_PLATFORM_TREE_MANAGER_H_
+  virtual AXPlat
\ No newline at end of file
DONE

To fix, run:

git apply <<DONE
diff --git a/impeller/scene/BUILD.gn b/impeller/scene/BUILD.gn
index 69c024f9e4..0000000000 100644
--- a/impeller/scene/BUILD.gn
+++ b/impeller/scene/BUILD.gn
@@ -43,15 +43,4 @@ impeller_component("scene") {
   deps = [ "//flutter/fml" ]
 }

-impeller_component("scene_unittests") {
-  testonly = true
-
-  sources = [ "scene_unittests.cc" ]
-
-  deps = [
-    ":scene",
-    "../fixtures",
-    "../playground:playground_test",
-    "//flutter/testing:testing_lib",
-  ]
-}
+impeller_component("sc
\ No newline at end of file
diff --git a/shell/platform/darwin/common/BUILD.gn b/shell/platform/darwin/common/BUILD.gn
index a3210bfed4..0000000000 100644
--- a/shell/platform/darwin/common/BUILD.gn
+++ b/shell/platform/darwin/common/BUILD.gn
@@ -35,20 +35,4 @@ source_set("framework_shared") {

   sources = [
     "framework/Source/FlutterChannels.mm",
-    "framework/Source/FlutterCodecs.mm",
-    "framework/Source/FlutterStandardCodec.mm",
-    "framework/Source/FlutterStandardCodecHelper.cc",
-    "framework/Source/FlutterStandardCodec_Internal.h",
-  ]
-
-  public = framework_shared_headers
-
-  defines = [ "FLUTTER_FRAMEWORK" ]
-
-  public_configs = [
-    "//flutter:config",
-    ":framework_relative_headers",
-  ]
-
-  deps = [ "//flutter/fml" ]
-}
+    "framework/S
\ No newline at end of file
DONE


stderr:
Performing C++/ObjC/Shader format check
Checking C++/ObjC/Shader formatting...
ERROR: Found 6 C++/ObjC/Shader files which were formatted incorrectly.
Found C++/ObjC/Shader format problems.
Performing GN format check
Checking GN formatting...
ERROR: Formatter command 'C:\Code\f\engine\src\flutter\third_party\gn\gn.exe format --stdin' failed with exit code 1. Command output follows:

ERROR at :28:14: Expecting assignment or function call.
  testonly = true
             ^---

ERROR: Found 2 GN files which were formatted incorrectly.
Found GN format problems.
Performing Java format check
Checking Java formatting...
ERROR: Cannot run Java, skipping Java file formatting!
Performing Python format check
Checking Python formatting...
All python files formatted correctly.
Performing Trailing whitespace format check
Checking for trailing whitespace on 336 source files...
No trailing whitespace found.

formatting checks finished in 0:00:04.654889
pre-push checks finished in 0:00:04.664406
error: failed to push some refs to 'github.com:loic-sharma/flutter-engine.git'
```

</details>

Would you be able to take a look at that? Those formatting recommendations seem incorrect. Attempting to format those same documents within Visual Studio does not produce any chances.

Also, I removed the test label :)

**Comment by mtolmacs** [comment]
This is indeed a breaking bug, the C++ formatting patch clearly cuts off the end of the file before diffing. Let me investigate this. In the meantime, this should not be merged.

Thanks for catching it!

**Comment by mtolmacs** [comment]
For reference, this is the dartlang issue associated with this bug: https://github.com/dart-lang/sdk/issues/50904

**Comment by Hixie** [comment]
@mtolmacs Is this PR still on your radar? Should we find a workaround for the Dart bug while we wait for it to be fixed?

**Comment by mtolmacs** [comment]
Hi @Hixie yes. It's functionally complete, the only thing needs fixing is the output issue this is blocked on currently. I don't have any other ideas how this can be worked around outside the Dart codebase economically, but I'm open to ideas.

**Comment by cbracken** [comment]
Added a teensy cleanup to make the linter happy.

File: ci/pubspec.yaml [hunk 4]
```diff
@@ -23,15 +23,27 @@ dependencies:
...
+  async_helper:
+    path: ../../third_party/dart/pkg/async_helper
   collection:
     path: ../../third_party/dart/third_party/pkg/collection
+  expect:
+    path: ../../third_party/dart/pkg/expect
   file:
     path: ../../third_party/pkg/file/packages/file
+  litetest:
+    path: ../testing/litetest
```
Comment (loic-sharma) [inline ci/pubspec.yaml:46]: Just to double-check, are the changes in this file necessary?

**Review by loic-sharma — APPROVED** [review]
Re-LGTM though please double-check the pubspec.yaml changes are necessary

**Comment by cbracken** [comment]
I've marked this as ready for review since the underlying blocker is now resolved. @mtolmacs can take one last pass to be sure he's good to go and reply re @loic-sharma's pubspec question above, and then let's land this thing 🚀.

File: ci/pubspec.yaml (hunk 4 above)
Comment (mtolmacs) [reply ci/pubspec.yaml:46]: These seem to be outdated, I cleaned them up. Thanks!

**Comment by mtolmacs** [comment]
I checked it with the latest Dart build from the tree (--no-prebuilt-dart-sdk) and removed the deprecated hack + tested it on Windows (I still need to test it on Mac and Linux locally), also adressing the pubspec.yaml question.

However the scripts hard-code the pre-built Dart SDK from the third-party directory, so we need to wait until the prebuild Dart SDK is updated to `3.1.0-189.0.dev`. I'm now sure how I can monitor that, so I will periodically check the tree and see if it's updated. Until then I'll keep testing it.

File: ci/pubspec.yaml (hunk 4 above)
Comment (mtolmacs) [reply ci/pubspec.yaml:45]: I had to re-add some of the dev dependencies because ci/test/format_test.dart uses them.

**Comment by loic-sharma** [comment]
> However the scripts hard-code the pre-built Dart SDK from the third-party directory, so we need to wait until the prebuild Dart SDK is updated to `3.1.0-189.0.dev`. I'm now sure how I can monitor that, so I will periodically check the tree and see if it's updated. Until then I'll keep testing it.

The Dart SDK in `//third_party/dart` should've been updated by https://github.com/flutter/engine/commit/ff7adb82f6169c7df424762a8440ff7ac71c9bc3. You should be able to get it locally using `gclient sync`. Please let me know if I'm missing something though!

**Comment by cbracken** [comment]
@mtolmacs you can build against the non-prebuilt Dart SDK by passing the `--no-prebuilt-dart-sdk` option to `flutter\tools\gn`. Sorry about the confusion.

@zanderso how frequently does the prebuilt Dart SDK roll in the engine? I can't remember if it's a CIPD package or something else. I had assumed it was updating roughly on the same frequency as the Dart SDK rolled into the engine.

**Comment by loic-sharma** [comment]
Ah interesting, it looks like we have two prebuilt Dart SDKs:

1. `//third_party/dart/tools/sdks/dart-sdk/bin/dart.exe`
2. `//flutter/prebuilts/windows-{x64,arm64}/dart-sdk/bin/dart.exe`

On my machine, `//flutter/prebuilts/windows-x64/dart-sdk/bin/dart.exe` reports:

```
Dart SDK version: 3.1.0-180.0.dev (dev) (Tue Jun 6 20:35:19 2023 -0700) on "windows_x64"
```

The [build uses `//flutter/prebuilts/` for its pre-built Dart SDK](https://github.com/flutter/engine/blob/5ff90e13818c9e388a30159f45445b07e060a887/tools/gn#L105-L132), it looks like we should update the [Windows](https://github.com/flutter/engine/blob/5ff90e13818c9e388a30159f45445b07e060a887/ci/format.bat#L23) and [non-Windows](https://github.com/flutter/engine/blob/5ff90e13818c9e388a30159f45445b07e060a887/ci/format.sh#L32) format scripts to reflect that.

(If we do need to update `//third_party/dart/tools/sdks/dart-sdk/bin/dart.exe`, it seems [these are the steps on the Dart repo](https://github.com/dart-lang/sdk/tree/main/tools/sdks#readme)).

**Comment by mtolmacs** [comment]
(sorry for the accidental self-rebase pushed, I was testing git push and was careless - reverted the commits in 664990c)

I've made the modification in `ci/format.bat` (and that file only, so not touching `format.sh`) to use the `//flutter/prebuilts/{platform}/dart-sdk` version of Dart. Tested it on Windows, so we can proceed if everyone approves.

Question regarding the new `ci/test/format_test.dart` file: Is there anything else needed here (for example hook this up with automated CI)? Currently it needs to be ran manually at the discretion of the developer.

**Comment by cbracken** [comment]
The engine tests are (mostly) centrally-managed in [run_tests.py](https://github.com/flutter/engine/blob/main/testing/run_tests.py) which is wired into CI in the build config files such as [this one](https://github.com/flutter/engine/blob/48e0b4e664222e81b5ace29623db8811ed19e7b3/ci/builders/mac_host_engine.json#L48-L62).

I think we're probably fine to land this as-is for the moment and land a followup patch that wires the new test into the python script.

**Comment by loic-sharma** [comment]
@mtolmacs FYI, I added your tests to the CI in [`150a5a7` (#36123)](https://github.com/flutter/engine/pull/36123/commits/150a5a7cd66a208fcf57d9cb5ce632b2b08f5d53). Let me know if you have feedback.

File: ci/bin/format.dart [hunk 5]
```diff
@@ -401,9 +395,23 @@ class ClangFormatChecker extends FormatChecker {
...
-          WorkerJob(<String>['diff', '-u', completedJob.command.last, '-'],
+          WorkerJob(<String>[
+            'git',
+            'diff',
+            '--no-index',
+            '--no-color',
+            '--ignore-cr-at-eol',
+            '--',
+            completedJob.command.last,
+            '-'
```
Comment (zanderso) [inline ci/bin/format.dart:406]:
```suggestion
            '-',
```

File: ci/bin/format.dart [hunk 6]
```diff
@@ -536,16 +548,30 @@ class JavaFormatChecker extends FormatChecker {
...
-            <String>['diff', '-u', completedJob.command.last, '-'],
+            <String>[
+              'git',
+              'diff',
+              '--no-index',
+              '--no-color',
+              '--ignore-cr-at-eol',
+              '--',
+              completedJob.command.last,
+              '-'
```
Comment (zanderso) [inline ci/bin/format.dart:565]:
```suggestion
              '-',
```

File: ci/bin/format.dart [hunk 7]
```diff
@@ -627,45 +657,87 @@ class GnFormatChecker extends FormatChecker {
...
+          cmd,
+          stdinRaw: codeUnitsAsStream(
+              File(path.join(repoDir.absolute.path, file)).readAsBytesSync()),
+          name: <String>[...cmd, file].join(' '),
+        );
+        jobs.add(job);
+      } else {
+        jobs.add(WorkerJob(<String>[...cmd, file],
+            name: <String>[...cmd, file].join(' ')));
+      }
```
Comment (zanderso) [inline ci/bin/format.dart:675]: Think positive:

```suggestion
      if (fixing) {
        jobs.add(WorkerJob(
          <String>[...cmd, file],
          name: <String>[...cmd, file].join(' '),
        ));
      } else {
        final WorkerJob job = WorkerJob(
          cmd,
          stdinRaw: codeUnitsAsStream(
            File(path.join(repoDir.absolute.path, file)).readAsBytesSync(),
          ),
          name: <String>[...cmd, file].join(' '),
        );
        jobs.add(job);
      }
```

**Comment by loic-sharma** [comment]
@mtolmacs, thank you for sticking through this and landing this! Wonderful work! 🥳
